File: 003_array_duplicate/array_duplicate.rb
def solution1(array)
  # Simple solution using a hashmap
  int_map = Hash.new(0)
  array.each do |i|
    int_map[i] += 1

    return i if int_map[i] == 2
  end

  return nil
end

if __FILE__ == $0
  test_array = [*1..1000000]
  test_array[25853] = 4

  puts solution1(test_array)
end

File: 003_array_duplicate/ArrayDuplicate.java
import java.util.HashMap;
import java.util.Map;

class ArrayDuplicate {
    /*
     * Simple solution using a hashmap
     */
    public static Integer solution1(int[] array) {
        Map<Integer, Integer> intMap = new HashMap<Integer, Integer>();
        for (int i : array) {
            Integer j = intMap.get(i);
            if (j == null) {
                intMap.put(i, 1);
            } else {
                intMap.put(i, ++j);
            }

            if (intMap.get(i) == 2) {
                return i;
            }
        }

        return null;
    }

    public static void main(String[] args) {
        int length = 1000000;
        int[] testArray = new int[length];
        for (int i = 1; i < length; i++) {
            testArray[i - 1] = i;
        }
        testArray[25853] = 4;
        System.out.println(solution1(testArray));
    }
}
